operator/internal/controller/pulumi/metrics_program_test.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumi
+
+import (
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
```

Comment:
I still don't like the linting sacrifice we take on by using these libraries. Especially in this case I think it makes things harder to read and more complicated than necessary. You're free to take this or leave it!

```go
package pulumi

import (
	"testing"

	"github.com/prometheus/client_golang/prometheus/testutil"
	"github.com/stretchr/testify/assert"
)

func TestProgramMetrics(t *testing.T) {
	numPrograms.Set(0)

	// it should increment the programs_active metric when a new Program is created
	newProgramCallback(nil)
	assert.Equal(t, 1.0, testutil.ToFloat64(numPrograms))

	// it should increment the programs_active metric when another Program is created
	newProgramCallback(nil)
	assert.Equal(t, 2.0, testutil.ToFloat64(numPrograms))

	// it should decrement the programs_active metric when a Program is deleted
	deleteProgramCallback(nil)
	assert.Equal(t, 1.0, testutil.ToFloat64(numPrograms))

	// it should decrement the programs_active metric when another Program is deleted
	deleteProgramCallback(nil)
	assert.Equal(t, 0.0, testutil.ToFloat64(numPrograms))

	// it should not decrement the programs_active metric when a Program is deleted and the metric is already at 0
	deleteProgramCallback(nil)
	assert.Equal(t, 0.0, testutil.ToFloat64(numPrograms))
}
```

Reply:
I thought about this, and in fact, I originally wrote this test using std Golang. However, I switched to using ginkgo ultimately, for consistency with other tests that we already have that test metrics. Also, this appeared to be more of a BDD style test, which is why I ultimately aligned with using Ginkgo.

Since we're already taking linting sacrifices in other files due to ginkgo, I'll leave this as is since I believe consistency here is more important.

Reply:
Yeah totally agree one port is the most reasonable. AFAICT the other port wasn't actually used in v1. I'm still wondering if it's OK to change the port, though. Existing service monitors would stop scraping metrics, right?